lib/appmap/hook.rb

```diff
@@ -120,25 +69,112 @@ def hook_builtins
           require hook.package.package_name if hook.package.package_name
           Array(hook.method_names).each do |method_name|
             method_name = method_name.to_sym
+            base_cls = class_from_string.(class_name)
 
-            cls = class_from_string.(class_name)
-            method = \
-              begin
-                cls.instance_method(method_name)
-              rescue NameError
-                cls.method(method_name) rescue nil
-              end
-
-            next if config.never_hook?(method)
+            hook_method = lambda do |entry|
+              cls, method = entry
+              return false if config.never_hook?(cls, method)
 
-            if method
               Hook::Method.new(hook.package, cls, method).activate
+            end
+
+            methods = []
+            methods << [ base_cls, base_cls.public_instance_method(method_name) ] rescue nil
+            if base_cls.respond_to?(:singleton_class)
+              methods << [ base_cls.singleton_class, base_cls.singleton_class.public_instance_method(method_name) ] rescue nil
+            end
+            methods.compact!
+            if methods.empty?
+              warn "Method #{method_name} not found on #{base_cls.name}"
             else
-              warn "Method #{method_name} not found on #{cls.name}"
+              methods.each(&hook_method)
             end
           end
         end
       end
     end
+
+    protected
+
+    def trace_class(trace_point)
+      path = trace_point.path
+
+      return if @notrace_paths.member?(path)
+
+      if config.path_enabled?(path)
+        location = trace_location(trace_point)
+        warn "Entering hook-enabled location #{location}" if Hook::LOG || Hook::LOG_HOOK
+        @trace_locations << location
+        unless @trace_end.enabled?
+          warn "Enabling hooking" if Hook::LOG || Hook::LOG_HOOK
+          @trace_end.enable
+        end
+      else
+        @notrace_paths << path
+      end
+    end
+
+    def trace_location(trace_point)
+      [ trace_point.path, trace_point.lineno ].join(':')
+    end
+
+    def trace_end(trace_point)
+      cls = trace_point.self
+
+      instance_methods = cls.public_instance_methods(false) - OBJECT_INSTANCE_METHODS
+      # NoMethodError: private method `singleton_class' called for Rack::MiniProfiler:Class
+      class_methods = begin
+        if cls.respond_to?(:singleton_class)
+          cls.singleton_class.public_instance_methods(false) - instance_methods - OBJECT_STATIC_METHODS
+        else
+          []
+        end
+      rescue NameError
+        []
+      end
+
+      hook = lambda do |hook_cls|
+        lambda do |method_id|
+          # Don't try and trace the AppMap methods or there will be
+          # a stack overflow in the defined hook method.
+          next if %w[Marshal AppMap ActiveSupport].member?((hook_cls&.name || '').split('::')[0])
```

Comment:
Seems like this comment needs to be updated: why are Marshal and ActiveSupport mentioned here?

Reply:
Well, same reason, the code becomes re-entrant.

Reply:
Ok. Has this always been true, or did something change that made it necessary to add them now? Is there a test that will break if the list needs to be updated again?